src/OrleansRuntime/GrainDirectory/RemoteGrainDirectory.cs

```diff
@@ -22,313 +20,52 @@ internal RemoteGrainDirectory(LocalGrainDirectory r, GrainId id)
-            }
-
-            if (router.MyAddress.Equals(owner))
-            {
-                router.RegistrationsSingleActLocal.Increment();
-                return partition.AddSingleActivation(address.Grain, address.Activation, address.Silo);
-            }
-
-            if (retries <= 0)
-                throw new OrleansException("Silo " + router.MyAddress + " is not the owner of the grain " +
-                                                address.Grain + " Owner=" + owner);
-
-            if (logger.IsVerbose2) logger.Verbose2("Retry {0} RemoteGrainDirectory.RegisterSingleActivation for address={1} at Owner={2}", retries, address, owner);
-            PrepareForRetry(retries);
-
-            await Task.Delay(RETRY_DELAY);
-
-            SiloAddress o = router.CalculateTargetSilo(address.Grain);
-            if (o == null)
-            {
-                // We don't know about any other silos, and we're stopping, so throw
-                throw new InvalidOperationException("Grain directory is stopping");
-            }
-            if (o.Equals(router.MyAddress))
-            {
-                router.RegistrationsSingleActLocal.Increment();
-                return partition.AddSingleActivation(address.Grain, address.Activation, address.Silo);
-            }
-            router.RegistrationsSingleActRemoteSent.Increment();
-            return await GetDirectoryReference(o).RegisterSingleActivation(address, retries - 1);
-        }
-
-        /// <summary>
-        /// Registers multiple new activations, in single activation mode, with the directory service.
-        /// If there is already an activation registered for any of the grains, then the corresponding new activation will
-        /// not be registered and the address of the existing activation will be returned.
-        /// Otherwise, the passed-in address will be returned.
-        /// <para>This method must be called from a scheduler thread.</para>
-        /// </summary>
-        /// <param name="addresses"></param>
-        /// <param name="retries">Number of retries to execute the method in case the virtual ring (servers) changes.</param>
-        /// <returns></returns>
-        public Task RegisterManySingleActivation(List<ActivationAddress> addresses, int retries)
+        public Task RegisterMany(List<ActivationAddress> addresses, bool singleActivation)
         {
             // validate that this request arrived correctly
             //logger.Assert(ErrorCode.Runtime_Error_100140, silo.Matches(router.MyAddress), "destination address != my address");
-            //var result = new List<ActivationAddress>();
-            if (logger.IsVerbose2) logger.Verbose2("RegisterManySingleActivation Count={0}", addresses.Count);
+
+            if (logger.IsVerbose2) logger.Verbose2("RegisterMany Count={0}", addresses.Count);
 
             if (addresses.Count == 0)
                 return TaskDone.Done;
```

Comment:
Only if you are sure that all callers check if the list is empty before calling this :)